### PR TITLE
add threshold of pattern match on systemlogmonitor

### PR DESCRIPTION
**Pull request by smileusd**
```release-note
This pr to add the feature on systemlogmonitor to trigger the condition only repeat match the rule pattern in a period.
For example I want add a pattern that only occur 3 times in 1 hour can trigger condition. This feature give a template for all systemlogmonitor to configure it.
If not configure them, that would be no change.
```

### CODE REVIEW

**Comment by smileusd** [comment]
@Random-Liu PTAL

**Comment by mmiranda96** [comment]
/ok-to-test

I will take a look at this PR next week, I'm a bit swamped with other work this one.

**Comment by smileusd** [comment]
@mmiranda96 Thank you trigger e2e. I will fix the test first.

File: pkg/systemlogmonitor/counts_buffer.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	return false
+}
+
+func NewCountBuffer(rules []systemlogtypes.Rule) map[string]*CountRingBuffer {
+	if len(rules) == 0 {
+		return nil
+	}
+	countBuffer := make(map[string]*CountRingBuffer)
+	for _, rule := range rules {
```
Comment (mmiranda96) [inline pkg/systemlogmonitor/counts_buffer.go:66]: Going with my suggested nested config, we can add a check here:
```go
if rule.CountThresholdConfig == nil {
  continue
}
```

File: pkg/systemlogmonitor/counts_buffer.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		length:       thd,
+		expirePeriod: ep,
+	}
+}
+
+func (crb *CountRingBuffer) IsMatchthreshold() bool {
+	tempIndex := crb.index
+	crb.index = (crb.index + 1) % crb.length
+    if crb.data[crb.index].IsZero() {
+		crb.data[tempIndex] = time.Now()
```
Comment (mmiranda96) [inline pkg/systemlogmonitor/counts_buffer.go:51]: Also, why are we using `Now()`. Shouldn't we use the time when the log entry appeared?

**Review by mmiranda96 — CHANGES_REQUESTED** [review]
Thanks for your contribution. There are several points on improvement on your PR, I have left a couple of comments. Also, could you explain more in detail how your implementation works? Please update this PR description with the details.

**Comment by smileusd** [comment]
@mmiranda96 PTAL

File: pkg/systemlogmonitor/counts_buffer.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+func NewCountRingBuffer(thd int, ep time.Duration) *CountRingBuffer {
+	return &CountRingBuffer{
+		data:         make([]time.Time, thd),
+		index:        0,
+		length:       thd,
+		expirePeriod: ep,
+	}
+}
+
+func (crb *CountRingBuffer) IsMatchthreshold() bool {
```
[the remark this answers is not in the document]
Comment (mmiranda96) [reply pkg/systemlogmonitor/counts_buffer.go:47]: Also, please rename to rename the function to `IsThresholdMatched`.

File: pkg/systemlogmonitor/counts_buffer.go (hunk 2 above)
Comment (mmiranda96) [reply pkg/systemlogmonitor/counts_buffer.go:51]: I'm not sure if the implementation actually depends on "now". But if it does, please consider using a clock instead, that way we can mock it in tests. See https://github.com/kubernetes/node-problem-detector/blob/fd51f17ec1f7f85086556c97b0d5efb40ead0fd6/pkg/systemlogmonitor/logwatchers/filelog/log_watcher.go#L65 for an example.

File: pkg/systemlogmonitor/counts_buffer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mmiranda96) [reply pkg/systemlogmonitor/counts_buffer.go:47]: Please add a comment explaining how this algorithm work.

**Review by mmiranda96 — CHANGES_REQUESTED** [review]
Please address the new and previous comments. Also, please go a bit more in depth on how the feature works, why it's needed, and what are the valid use cases.

**Pull request closed without merging**